### PR TITLE
Add "base", "final" and "interface" modifiers

### PR DESCRIPTION
**Pull request by brianquinlan**
Fixes https://github.com/dart-lang/http/issues/918

### CODE REVIEW

File: pkgs/http/lib/retry.dart [hunk 1]
```diff
@@ -14,7 +14,7 @@ import 'http.dart';
 /// NOTE: [RetryClient] makes a copy of the request data in order to support
 /// resending it. This can cause a lot of memory usage when sending a large
 /// [StreamedRequest].
-class RetryClient extends BaseClient {
+final class RetryClient extends BaseClient {
```
Comment (brianquinlan) [inline pkgs/http/lib/retry.dart:17]: Interface is identical to client. There is no reason to implement or extend this.

File: pkgs/http/lib/src/browser_client.dart [hunk 2]
```diff
@@ -31,7 +31,7 @@ BaseClient createClient() {
 /// [BaseRequest.followRedirects], and [BaseRequest.maxRedirects] fields. It is
 /// also unable to stream requests or responses; a request will only be sent and
 /// a response will only be returned once all the data is available.
-class BrowserClient extends BaseClient {
+base class BrowserClient extends BaseClient {
```
Comment (brianquinlan) [inline pkgs/http/lib/src/browser_client.dart:34]: Interface is identical to Client

File: pkgs/http/lib/src/io_client.dart [hunk 3]
```diff
@@ -43,7 +43,7 @@ class _ClientSocketException extends ClientException
 }
 
 /// A `dart:io`-based HTTP client.
-class IOClient extends BaseClient {
+base class IOClient extends BaseClient {
```
Comment (brianquinlan) [inline pkgs/http/lib/src/io_client.dart:46]: Interface is identical to Client

File: pkgs/http/test/io/client_test.dart [hunk 4]
```diff
@@ -14,14 +14,14 @@ import 'package:test/test.dart';
 
 import '../utils.dart';
 
-class TestClient extends http.BaseClient {
+base class TestClient extends http.BaseClient {
```
Comment (brianquinlan) [inline pkgs/http/test/io/client_test.dart:17]: Interface is identical to Client

File: pkgs/http/test/io/http_test.dart [hunk 5]
```diff
@@ -10,7 +10,7 @@ import 'package:test/test.dart';
 
 import '../utils.dart';
 
-class TestClient extends http.BaseClient {
+base class TestClient extends http.BaseClient {
```
Comment (brianquinlan) [inline pkgs/http/test/io/http_test.dart:13]: Interface is identical to Client

**Comment by brianquinlan** [comment]
I didn't add `base` to `BaseRequest`, `BaseClient` and `BaseResponse` but someone might want to `implement` (e.g. for mocking) an implementor of those classes.

File: pkgs/http/lib/src/browser_client.dart (hunk 2 above)
Comment (natebosch) [reply pkgs/http/lib/src/browser_client.dart:34]: I wonder if this should be `final`?

Comment (brianquinlan) [reply pkgs/http/lib/src/browser_client.dart:34]: I could imagine people wanting to add methods to `Client` implementations. Like:

````dart
class DownloadingBrowserClient extends BrowserClient {
  Stream<UInt8List> downloadFile(Url) {
     ...
  }

}```

**Review by natebosch — APPROVED** [review]

File: pkgs/http/test/io/client_test.dart [hunk 6]
```diff
@@ -14,14 +14,14 @@ import 'package:test/test.dart';
...
   @override
   Future<http.StreamedResponse> send(http.BaseRequest request) {
     throw UnimplementedError();
   }
 }
 
-class TestClient2 extends http.BaseClient {
+base class TestClient2 extends http.BaseClient {
```
[the remark this answers is not in the document]
Comment (lrhn) [reply pkgs/http/test/io/client_test.dart:24]: Ditto. Just no modifier. It's not a public class, so we don't care what people can do with it.

File: pkgs/http/lib/src/byte_stream.dart [hunk 7]
```diff
@@ -7,7 +7,7 @@ import 'dart:convert';
 import 'dart:typed_data';
 
 /// A stream of chunks of bytes representing a single piece of data.
-class ByteStream extends StreamView<List<int>> {
+final class ByteStream extends StreamView<List<int>> {
```
Comment (lrhn) [inline pkgs/http/lib/src/byte_stream.dart:10]: (This class should not need to exist. It should just be extension methods on `Stream<List<int>>`.)

**Review by lrhn — APPROVED** [review]
The `BaseClient` class is documented as "This is a mixin-style class" and it can be used as a mixin today, so I'd recommend making it a `mixin class`.

LGTM without the `base`s.

(The package could generally use some clean-up. A lot of the code seems overly complicated, and could use more modern features or just be optimized better.)

**Pull request merged**